google/ggrc-core
# GGRC-5522 Avoid deadlocks on ACL propagation

# Issue description

2+ concurrent requests enter a deadlock while performing propagation.

# Steps to test the changes

Run imports of 1000 assessments for the same Audit in 3-4 parallel jobs.

Expected result: the Assessments are created, their ACL fields are imported, the Program and Audit roles are propagated to the new Assessments.
Actual result: the Assessments are created, but either ACL fields or propagated roles are missing; multiple "import failed due to unknown error" are reported on row level, and some of the jobs are reported crashed.

# Solution description

Multiple solutions were attempted:
- lower transaction isolation right before doing the INSERT FROM SELECT to avoid locks from the SELECT part (doesn't work by design for transactions that already performed at least one SQL statment)
- lower transaction isolation in the hooked commit before performing ACL propagation to avoid locks from the SELECT part of INSERT FROM SELECT (multiple tests fail with this change; apparently, lower transaction isolation affects other transactions as well, no detailed analysis was performed)
- perform SELECT and INSERT separately (the current solution; plain SELECTs don't set any locks on any tables).

Since a big amount of rows are propagated for bigger imports (285000 rows was reported in an error log), MySQL closes the connection while receiving the big data packet, therefore we should insert the data chunk by chunk. Chunk size of 1000 was chosen as an arbitrary low limit.

For some reason, if we don't COMMIT right after doing SELECT for the whole data set, deadlocks still occur. It is possible that MySQL sets some read locks when doing a SELECT with multiple JOINs, but no detailed analysis was done in this field either.

# Results of testing on ggrc-perf

Testing scenario: each import creates 1000 Assessments on Audit 4567 (a big one) with Assessment Template having 10 custom attributes. The imported CSVs are the same for each job.


## Before the fix

1 out of 4 import jobs succeeds, occasionally all 4 jobs fail.

## After the fix

### With chunk size of 1000

4 out of 4 import jobs succeed, they take 5, 5, 5, 13 hours each.

### With chunk size of 50000:

1. 4 out of 5 import jobs succeeded, one failed with "Lost connection to MySQL server" in an unrelated part of code (some SELECT to issuetracker_issues).
2. 4 out of 6 import jobs succeeded, one failed with "Lost connection to MySQL server" on a hook that invalidates changes to built-in roles (unrelated to my changes), another failed because a person stopped the app before the imports were finished.
3. 4 out of 5 import jobs succeeded, one failed because the worker was killed without any error message in the logs (definitely not caused by memory consumption or timeouts, judging by the logs).

Note: it appears that the amount of data affects the import time. Trial 1 was done on an Audit with 2000 Assessments, and the DB wasn't cleaned since then. The timings for successful jobs are (in seconds):
1. 6973, 4952, 5023, 5995
2. 14069, 11093, 13114, 12986
3. 22796, 22240, 21129. 19352

Since the code wasn't changed between the runs, the reason of increased timings is data.

# Sanity checklist

- [ ] My changes follow our [performance guidelines](https://github.com/google/ggrc-core/blob/dev/docs/source/contributing/performance.rst). - a big dataset is fetched from the DB and pushed back again in multiple requests, which doesn't feel a performant job



# PR Review checklist
